src/rapids_metadata/metadata.py

```diff
@@ -12,36 +12,54 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from dataclasses import dataclass, field
+from enum import Enum
+from os import PathLike
 from typing import Union
 
+from .rapids_version import get_rapids_version
 
-class Metadata:
-    def __init__(self, inherit: Union["Metadata", None]=None):
-        self._packages: set[str] = set()
-        self._non_alpha_spec_packages: set[str] = set()
-        self._non_cuda_suffixed_packages: set[str] = set()
+__all__ = ["PseudoRepository", "RAPIDSMetadata", "RAPIDSPackage", "RAPIDSRepository", "RAPIDSVersion"]
 
-        if inherit:
-            self._packages.update(inherit.packages)
-            self._non_alpha_spec_packages.update(inherit.non_alpha_spec_packages)
-            self._non_cuda_suffixed_packages.update(inherit.non_cuda_suffixed_packages)
 
-    @property
-    def packages(self) -> set[str]:
-        return self._packages
+class PseudoRepository(Enum):
+    NVIDIA = "nvidia"
 
-    @property
-    def non_alpha_spec_packages(self) -> set[str]:
-        return self._non_alpha_spec_packages
+    def __str__(self):
+        return f"_{self.value}"
+
+
+@dataclass
+class RAPIDSPackage:
+    has_alpha_spec: bool = field(default=True)
+    has_cuda_suffix: bool = field(default=True)
+
+
+@dataclass
+class RAPIDSRepository:
+    packages: dict[str, RAPIDSPackage] = field(default_factory=dict)
+
+
+@dataclass
+class RAPIDSVersion:
+    repositories: dict[Union[str, PseudoRepository], RAPIDSRepository] = field(default_factory=dict)
 
     @property
-    def non_cuda_suffixed_packages(self) -> set[str]:
-        return self._non_cuda_suffixed_packages
+    def all_packages(self):
+        return {package for _, repository_data in self.repositories.items() for package in repository_data.packages}
 
     @property
-    def alpha_spec_packages(self) -> frozenset[str]:
-        return frozenset(self.packages - self.non_alpha_spec_packages)
+    def alpha_spec_packages(self):
+        return {package for _, repository_data in self.repositories.items() for package, package_data in repository_data.packages.items() if package_data.has_alpha_spec}
 
     @property
-    def cuda_suffixed_packages(self) -> frozenset[str]:
-        return frozenset(self.packages - self.non_cuda_suffixed_packages)
+    def cuda_suffixed_packages(self):
+        return {package for _, repository_data in self.repositories.items() for package, package_data in repository_data.packages.items() if package_data.has_cuda_suffix}
+
+
+@dataclass
+class RAPIDSMetadata:
+    versions: dict[str, RAPIDSVersion] = field(default_factory=dict)
+
+    def get_current_version(self, directory: PathLike=None) -> RAPIDSVersion:
```

Comment:
The vast majority of the time, this is going to be called from a process with a current directory that has `VERSION` in it or above it. That's the primary intended use case. Frankly the `directory` argument is extra. I'm open to the idea of splitting these two into different functions.

Reply:
Even with that assumption, I think it would be both safer and clearer for, for example, the code in `pre-commit-hooks` that'll use this to call it like this:

```python
rapids_metadata.get_current_version(
    directory=os.getcwd()
)
```

instead of like this:

```python
rapids_metadata.get_current_version()
```

That makes the assumptions about working directory a bit more explicit.

BUT... I don't feel strongly enough about this particular thing to block the PR. I'd prefer leaving the method totally unchanged to splitting it up into one that requires directory and one that doesn't.

Or... maybe just removing the `directory` argument altogether, and adding it to the interface later, when a specific use case for it shows up?

Reply:
I decided to remove the default value from `directory`. We can just use `os.getcwd()` or `"."` for now.